[RFC] Ephemeral resources and write-only attributes (#2793)
opentofu/opentofu: rfc/20250317-ephemeral-resources.md

# Ephemeral resources, variables, outputs, locals and write-only arguments

Issue: https://github.com/opentofu/opentofu/issues/1996

Right now, OpenTofu information for resources and outputs are written to state as it is. This is presenting a security risk

> as some of the information from the stored objects can contain sensitive bits that can become visible to whoever is having access to the state file.

Reviewer: It also might be worth introducing the idea that state encryption is our currently recommended solution.

IMO they are two overlapping, but not exclusive concepts

Advantages of State Encryption:
* All secrets are protected, regardless of provider or resource/module configuration
* Don't need to chose what is protected and what is exposed
* Also protects against provider injection
  - https://www.plerion.com/blog/hacking-terraform-state-for-privilege-escalation
  
  Advantages of Ephemeral
  * No keys to loose / have leaked, secrets (if identified properly) are not stored in the first place.

> as some of the information from the stored objects can contain sensitive bits that can become visible to whoever is having access to the state file.

Reviewer 2: That raises another question, we should explain how we shouldn't use this with state encryption (due to the ephemerality) because some people may think "This is great for secrets, i can use it for my passphrase" and have issues.

> as some of the information from the stored objects can contain sensitive bits that can become visible to whoever is having access to the state file.

Author: Please check this out: ae712fdaf7a5ba4b69cfcf0e65d960df7d301dac

> In order to provide a better solution for the aforementioned situation, OpenTofu introduces the concept of "ephemerality".

Author: Please check this out ae712fdaf7a5ba4b69cfcf0e65d960df7d301dac

Any new feature under this new concept will provide ways to skip values from being written to the state and plan files.

To make this work seamlessly with most of the blocks that OpenTofu supports, the following functionalities need to be able to work with the ephemeral concept:
* `resource`'s `write-only` attributes
* variables
* outputs
* locals

> * `ephemeral` resources

Reviewer 2: I just want to go on record that I don't like how they're called `ephemeral resources` because this is confusing. Mainly because `resource` and `ephemeral` here are 2 different types of blocks.

I would prefer a different name, such as "ephemeral entities" etc, but I fear it may be too late.

Author: I don't think that it's too late and I agree that it can become confusing. If others agree with this, it's an easy change:
* "%s/ephemeral resource/ephemeral entity/g"
* "%s/ephemeral resources/ephemeral entities/g"

Hitting enter? 😅

Therefore, these attributes are suitable for configuring specific resources with sensitive data, like passwords, access keys, etc.

A write-only attribute can accept an ephemeral or a non-ephemeral value, even though it's recommended to use ephemeral values for such attributes.

Because these attributes are not written to the plan file, the update of a write-only attribute it's getting a little bit trickier.
Provider implementations do generally include also a "version" argument linked to the write-only one.
For example having a write-only argument called `secret`, providers should also include
a non-write-only argument called `secret_version`. Every time the user wants to update the value of `secret`, it needs to change the value of `secret_version` to trigger a change.
The provider implementation is responsible with handling this particular case: because the version attribute is stored also in the state, the provider needs to compare the value from the state with the one from the configuration and in case it differs, it will trigger the update of the `secret` attribute.

> The write-only attributes are supported momentarily by a low number of providers and resources.

Author: In 91f105241ac49c3c20ba51d52306be62c7ac4c03

Having the `aws_db_instance` as one of those, here is an example on how to use the write-only attributes:
```hcl
resource "aws_db_instance" "example" {
  // ...
  password_wo         = "your-initial-password"
  password_wo_version = 1
  // ...
}
```
By updating **only** the `password_wo`, on the `tofu apply`, the password will not be updated.
To do so, the `password_wo_version` needs to be incremented too:
```hcl
resource "aws_db_instance" "example" {
  // ...
  password_wo         = "new-password"
  password_wo_version = 2
  // ...
}
```

As seen in this particular change of the [terraform-plugin-framework](https://github.com/hashicorp/terraform-plugin-framework/commit/ecd80f67daed0b92b243ae59bb1ee2077f8077c7), the write-only attribute cannot be configured for set attributes, set nested attributes and set nested blocks.

> Write-only attributes cannot generate a plan diff because the prior state does not contain a value that OpenTofu can use to compare the new value against and also the planned value of a write-only argument will always be empty.

Author: b4121ce87a6c902a783827039b5f38ddac6d7bc6

### Variables
Any `variable` block can be marked as ephemeral.
```hcl
variable "ephemeral_var" {
  type      = string
  ephemeral = true
}
```
OpenTofu will allow usage of these variables only in other ephemeral contexts:
* write-only arguments
* other ephemeral variables
* ephemeral outputs
* local values
* ephemeral resources
* provisioner blocks
* connection blocks
* provider configuration

> Usage in any other place will raise an error:

Author: Changed in 119b5c9a3c32f5a41a671586e4852515b4ee3b34. Indeed there were a lot of places. Sorry for this, seems a bias on my side. Weird.

```shell
│ Error: Invalid use of an ephemeral value
│
│   with playground_secret.store_secret,
│   on main.tf line 30, in resource "playground_secret" "store_secret":
│   30:   secret_name = var.password
│
│ "secret_name" cannot accept an ephemeral value because it is not a write-only attribute which means that will be written to the state.
╵
```

OpenTofu will not store ephemeral variable(s) in plan files. 
If a plan is generated from a configuration that is having at least one ephemeral variable, 
when the planfile will be applied, the value(s) for the ephemeral variable(s) needs to passed again by 

> using `-var` or `-var-file` arguments.

Author: c47bbb084a80d34a729715b39653b14bc870f3e5

### Outputs
An `output` block can be configured as ephemeral as long as it's
not from the root module. 

> This limitation is natural since ephemeral outputs are meant to be skipped from the state file. Therefore, there is no usage of such a defined output block in a root module.

Author: 603979cec2bf7e3c76064ef7515982f3c5d1c703

> Ephemeral outputs are useful when a child module returns sensitive data, allowing the caller to use the value of that output in other ephemeral contexts.

Author: 14bc104ef981f19fe02640e00ee25e76ccab2070

To mark an output as ephemeral, use the following syntax:
```hcl
output "test" {
  // ...
  ephemeral = true
}
```

The ephemeral outputs are available during plan and apply phase and can be accessed only in specific contexts:
* ephemeral variables
* other ephemeral outputs
* write-only attributes
* ephemeral resources
* `provisioner` block
* `connection` block

### Locals
Local values are automatically marked as ephemeral if any of the value that is used to compute the local is already an ephemeral one.

Eg:
```hcl
variable "a" {
  type = string
  default = "a value"
}

variable "b" {
  type = string
  default = "b value"
  ephemeral = true
}

locals {
  a_and_b = "${var.a}_${var.b}"
}
```
Because variable `b` is marked as `ephemeral`, then the local `a_and_b` is marked as `ephemeral` too.

Locals marked as ephemeral are available during plan and apply phase and can be referenced only in specific contexts:
* ephemeral variables
* other ephemeral locals
* write-only attributes
* ephemeral resources
* `provider` blocks configuration
* `connection` and `provisioner` blocks

### Ephemeral resource

> In contrast with the write-only arguments where only specifically tagged attributes are skipped from the state/plan file, `ephemeral` resources are skipped entirely.

Author: e7c684a0fbb6f68fd87bcaeec74e17c7978d76bb

  #   command = "echo write-only value: ${aws_secretsmanager_secret_version.store_from_ephemeral_output.secret_string_wo}"
  # }

  provisioner "local-exec" { # (10)
    when    = create
    command = "echo ephemeral value from module: #${jsonencode(module.secret_management.secrets)}#"
  }

  provisioner "local-exec" {
    when    = create
    command = "echo non-ephemeral variable from root: #${jsonencode(var.secrets)}#"
  }
}
```

- (1) Variables will be able to be marked as ephemeral. By doing so, those will be able to be used only in ephemeral contexts.
- (2) Version field that is going together with the actual write-only argument to be able to update the value of it. To upgrade the secret, the version field needs to be updated, otherwise OpenTofu will generate no diff for it.
- (3) Using ephemeral resource to retrieve the secret. Maybe looks a little bit weird, because right above we are having the resource of the same type that is looking like it should be able to be used to get the secret. In reality, because that `resource` is using `secret_string_wo` to store the information, that field is going to be null when referenced. Check (9) for more details.

> - (4) Module output that is referencing an ephemeral value, it needs to be marked as ephemeral too. Otherwise, OpenTofu will generate an error.

Author: 4d02f23b7bca34f689e30afa7290ed9cdd355c7d

- (5) The variable that is going to be used in an ephemeral variable, is not required to be ephemeral. The value can also be a hardcoded value without being ephemeral.
- (6) Here we used another `aws_secretsmanager_secret` just to have an easier example on how ephemeral/write-only/(ephemeral variables)/(ephemeral outputs) are working together. But there will be more and more resources that will allow to work with this new concept.
- (7) Referencing a module ephemeral output to ensure that the ephemeral information is passed correctly between two modules.

> - (8) Creating an output that is referencing a write-only argument *requires* the output to be marked as sensitive. 

Author: Changed here: 88a7826941e9048742cd732405ac2cf6f3417da6
I agree that would make more sense on having this also for the child modules, but do we want to create this diff between TF and OpenTofu?

> - (8) Creating an output that is referencing a write-only argument *requires* the output to be marked as sensitive. 

Reviewer 2: I think we should consider if it's useful or not. I don't have much of an opinion either way here really. I feel like root level only is fine.

> - (8) Creating an output that is referencing a write-only argument *requires* the output to be marked as sensitive. 

Author: Follow up on this: After further investigation on this one, this was having **nothing** to do with write-only argument being referenced in an output. The provider field I was using, was marked as `sensitive` in the provider schema. Therefore, [this](https://github.com/opentofu/opentofu/blob/d1f0999aed2b171f62ae802c23b9674ca0afda6c/internal/tofu/node_output.go#L377-L389) is the code that was executed.
Still, the rule with only the root output still stands but it's in OpenTofu for a long time now.

Apply complete! Resources: 5 added, 0 changed, 0 destroyed.
```

Breakdown:
- (11) - If an ephemeral block is referencing any unknown value, the opening is deferred for later, when the value will be known.
- (12).  As can be seen, the ephemeral resources are not shown in the list of changes. The only mention of those is in the actual action logs where we can see that it opens and closing those.
- (13).  This will be visible in the action logs, while an ephemeral resource will be opened.
- (14).  This is how a provisioner output should look like when an ephemeral value is used inside.
- (15).  This will be visible in the action logs, while an ephemeral resource will be closed.

## Technical Approach
In this section, as in the "Proposed Solution" section, we'll go over each concept, but this time with a more technical focus.

### Write-only arguments
Most of the write-only arguments logic is already in the [provider-framework](https://github.com/hashicorp/terraform-plugin-framework):
* [Initial implementation](https://github.com/hashicorp/terraform-plugin-framework/pull/1044)

> * [Sets comparisson enhancement](https://github.com/hashicorp/terraform-plugin-framework/pull/1064)

Author: 11730c4361e2d3137791ceeb79a6a3cce5444d44

  * Test this in-depth for all the block types except sets of any kind (Investigate and understand why sets are not allowed by the plugin framework).
    * Add a new validation on the provider schema to check against, set nested attributes and set nested blocks with writeOnly=true. Tested this with a version of terraform-plugin-framework that allowed writeOnly on sets and there is an error returned. (set attributes are allowed based on my tests)
    In order to understand this better, maybe we should allow this for the moment and test OpenTofu with the [plugin-framework version](https://github.com/hashicorp/terraform-plugin-framework/commit/0724df105602e6b6676e201b7c0c5e1d187df990) that allows sets to be write-only=true.

> [!NOTE]
>
> Write-only attributes will be presented in the OpenTofu's UI as `(write-only attribute)` instead of the actual value.

### Variables

For enabling ephemeral variables, these are the basic steps that need to be taken:
* Update config to support the `ephemeral` attribute.

> * Mark the variables with a new mark ensure that the marks are propagated correctly.

Author: I am not aware of this. Could you expand it a little bit please?

> * Mark the variables with a new mark ensure that the marks are propagated correctly.

Reviewer 2: https://github.com/opentofu/opentofu/pull/2503

> * Mark the variables with a new mark ensure that the marks are propagated correctly.

Author: Oh, this is what you meant! Ok. I misread your initial comment.
Updated in 2695d6c58dc739ceae192242733ddea987cdf450

* Based on the marks, ensure that the variable cannot be used in other contexts than the ephemeral ones (see the User Documentation section for more details on where this is allowed).
* Check the state of [#1998](https://github.com/opentofu/opentofu/pull/1998). If that is merged, in the changes where variables from plan are verified against the configuration ones, we also need to add a validation on the ephemerality of variables. If the variable is marked as ephemeral, then the plan value is allowed (expected) to be missing. 
* Ensure that when the prompt is shown for an ephemeral variable there is an indication of that:
  ```hcl
  var.password (ephemeral)
     Enter a value:

>   ````

Author: Thanks. Fixed in 4d02f23b7bca34f689e30afa7290ed9cdd355c7d

* If a module is having an ephemeral variable declared, that variable can get values from any source, even another non-ephemeral variable.

> We should use boolean marks, as no additional information is required to be carried. When introducing the marks for these, extra care should be taken in *all* the places marks are handled and ensure that the existing implementation around marks is not affected.

Reviewer 2: do we also need to introduce some checks/tests to ensure that functions such as `sensitive()` do not accidentally strip the ephemeral mark too?

Author: 3fe9247caedeef0099138111dece7a18065cda2b

### Outputs

For enabling ephemeral outputs, these are the basic steps that need to be taken:
* Update config to support the `ephemeral` attribute.
* Mark the outputs with a new mark and ensure that the marks are propagated correctly.
  * We should use boolean marks, as no additional information is required to be carried. When introducing the marks for these, extra care should be taken in *all* the places marks are handled and ensure that the existing implementation around marks is not affected.
* Based on the marks, ensure that the output cannot be used in other contexts than the ephemeral ones (see the User Documentation section for more details on where this is allowed).

> [!TIP]
>
> For an example on how to properly introduce a new mark in the outputs, you can check the [PR](https://github.com/opentofu/opentofu/pull/2633) for the deprecated outputs.

Strict rules:

> * A root module cannot define ephemeral outputs since are useless.

Reviewer 2: ```suggestion
* A root module cannot define ephemeral outputs.
```

Author: ce97918f1ad6429e17d88f7c250530742926f3f0

    │     sensitive = true
  ```

Considering the rules above, root modules cannot have any ephemeral outputs defined.

### Locals
Any `local` declaration will be marked as ephemeral if in the expression that initialises it an ephemeral value is used:
```hcl
variable "var1" {
  type = string
}

variable "var2" {
  type = string
}

variable "var3" {
  type      = string
  ephemeral = true
}

locals {
  eg1 = var.var1 == "" ? var.var2 : var.var1 // not ephemeral
  eg2 = var.var2 // not ephemeral
  eg3 = var.var3 == "" ? var.var2 : var.var1 // ephemeral because of var3 conditional
  eg4 = var.var1 == "" ? var.var2 : var.var3 // ephemeral because of var3 usage
  eg5 = "${var.var3}-${var.var1}" // ephemeral because of var3 usage
  eg6 = local.eg4 // ephemeral because of eg4 is ephemeral
}
```

Once a local is marked as ephemeral, this can be used only in other ephemeral contexts. Check the `User Documentation` section for more details on the allowed contexts.

> 

Reviewer 2: let's not forget to document here that variables that take in ephemeral marked values should also be implicitly marked as ephemeral

Author: Added information in 11a29cb050fd95ab55b5f6d5df01bba3c73af854.
A variable will not become ephemeral strictly from referencing an ephemeral value. In order for it to be able to work with ephemeral values, it needs to be configured specifically.

### Ephemeral resources
Due to the fact ephemeral resources are not stored in the state/plan file, this block is not creating a diff in the OpenTofu's UI.

> Instead, OpenTofu will notify the user of opening/renewing/closing an ephemeral resource with messages similar to the following:

Reviewer 2: Can we discuss what these actions mean? `opening` `renewing` and `closing`?  As someone approaching this fresh I think I can assume what it means but I think we should declare it explicitly so we can have a conversation about it. 

It feels like a strange set of verbs to show up in the ui to me (Opening.. Closing... etc)

Author: Changed a little in 24b57e7c7b16816ae8e5d98b57658781656179c9.
But please check the specific chapter dedicated to the `Renew` [method](https://github.com/opentofu/opentofu/blob/6b81cb76c65cb76311db9ecf9d67a94fbb0185cc/rfc/20250317-ephemeral-resources.md#renew-method-details).

```bash
ephemeral.playground_random.password: Opening...
ephemeral.playground_random.password: Opening succeeded after 0s
ephemeral.playground_random.password: Closing...
ephemeral.playground_random.password: Closing succeeded after 0s
```

Ephemeral resources lifecycle is similar with the data blocks:
* Both basic implementations require the same methods (`Metadata` and `Schema`) while datasource is defining `Read` compared with the ephemeral resource that is defining `Open`. When talking about the basic functionality of the ephemeral resources, the `Read` method will behave similarly with the `Read` on a datasource, where it reads the data.
* Also, both blocks support `Configure`, `ConfigValidators` and `ValidateConfig` as extensions of the basic definition.
* Ephemeral resources do support two more operations in contrast with datasources:
  * `Renew`

>     * Together with the data returned by the `Open` method call, the provider can also specify a `RenewAt` which will be a specific moment in time when OpenTofu should call the `Renew` method to get an updated information from the ephemeral resource. OpenTofu will have to check for `RenewAt` value anytime it intends to use the value returned by the ephemeral resource.

Reviewer 2: This renewat functionality isn't super clear to me based on the doc above, I wasn't expecting this. 

Can you describe it in abit more detail and why it's needed? Is it so that providers can determine if/when opentofu should refetch rather than it only being a human (version) driven re-fetch?

Author: Is the `Renew` [method] (https://github.com/opentofu/opentofu/blob/6b81cb76c65cb76311db9ecf9d67a94fbb0185cc/rfc/20250317-ephemeral-resources.md#renew-method-details) dedicated section enough?

  * `Close`

>     * When an ephemeral resource is having this method defined, it is expecting it to be called in order to release a possible held resource. A good example of this is with a Vault provider that could provide a secret by obtaining a lease, and when the secret is done being used, OpenTofu should call `Close` on that ephemeral resource to instruct on releasing the lease and revoking the secret.

Reviewer 2: ```suggestion
    * When an ephemeral resource is having this method defined, it is expecting it to be called in order to release a possible held resource. A good example of this is with a Vault/OpenBao provider that could provide a secret by obtaining a lease, and when the secret is done being used, OpenTofu should call `Close` on that ephemeral resource to instruct on releasing the lease and revoking the secret.
```

Author: da777bb65a5a2162f48c80bca72301d6412b937f

>     * When an ephemeral resource is having this method defined, it is expecting it to be called in order to release a possible held resource. A good example of this is with a Vault provider that could provide a secret by obtaining a lease, and when the secret is done being used, OpenTofu should call `Close` on that ephemeral resource to instruct on releasing the lease and revoking the secret.

Reviewer 2: >it to be called in order to release a possible held resource.

When, how?

Author: Added in https://github.com/opentofu/opentofu/commit/24b57e7c7b16816ae8e5d98b57658781656179c9#diff-1b793b731c0b3e0929632ddfa87429e5d0d2ddbe95bd3f9a09dea465ac1d406dR587-R597

Do you think that we should go more in-depth? Most probably the work of this should be done in `graphNodeCloseProvider`, but I think that this kind of details should be left for the implementor to decide.
Please correct me if I am wrong here.

* lifecycle
* count
* for_each
* depends_on
* provider

#### `Open` method details
When OpenTofu will have to use an ephemeral resource, it needs to call its `Open` method, passing over the config of the ephemeral resource.

The call to the `Open` method will return the following data:
* `Private` that OpenTofu is not going to use in other contexts than calling the provider `Close` or `Renew` optionally defined methods.
* `Result` that will contain the actual ephemeral information. This is what OpenTofu needs to handle to make it available to other ephemeral contexts to reference.
* `RenewAt` being an optional timestamp indicating when OpenTofu will have to call `Renew` method on the provider before using again the data from the `Result`.

Observations:
* In the `Result`, OpenTofu is epecting to find any non-computed given values in the request, otherwise will return an error.
* In the `Result`, the fields marked as computed can be either null or have an actual value. If an unknown if found, OpenTofu will return an error.

> [!NOTE]
>

> > If any information in the configuration of an ephemeral resource is unknown during the `plan` phase, OpenTofu will defer the provisioning of the resource for the `apply` phase.

Reviewer 2: ```suggestion
> If any information in the configuration of an ephemeral resource is unknown during the `plan` phase, OpenTofu will defer the provisioning of the resource for the `apply` phase. This means that inconsistency can occur between the plan and the apply phase.
```

Author: f603db886aa3d830534b5cc68f7e415e58e3d343

#### `Renew` method details
The `Renew` method is called only if the response from `Open` or another `Renew` call is containing a `RenewAt`.
When `RenewAt` is present, OpenTofu, before using the `Result` from the `Open` method response, will check if the current timestamp is at or over `RenewAt` and will call the `Renew` method by providing the previously returned `Private` information.

> [!NOTE]
> 
> `Renew` does not return a *new* information meant to replace the initial `Result` returned by the `Open` call.

> > Due to this, `Renew` is only useful for system similar to Vault where the lease can be renewed without generating new data.

Reviewer 2: ```suggestion
> Due to this, `Renew` is only useful for systems where an entity can be renewed without generating new data.
```

Author: f603db886aa3d830534b5cc68f7e415e58e3d343

#### `Close` method details

> When OpenTofu is done using an ephemeral resource, it needs to call its `Close` method to ensure that any remote data associated with the data returned in `OpenResponse.Result` is released and/or cleaned up properly.

Reviewer 2: When you say "Done using", is that
- when the apply phase is finished
- when the value is no longer referenced by a resource/output?
- when the node in the graph has been read completely

Can you elaborate on what `done` is here please?

Author: Yeah, I am having a rough time wording that correctly. "done using" in this context means that the apply phase finished and now it executes the close nodes. Right before executing `graphNodeCloseProvider` nodes, it should call `Close` on the ephemeral resources.
If you can help with a good wording around this, would greatly appreciate it.

> The `Renew` request is requiring the latest `Private` data returned by the call to `Open` or `Renew` method.

Reviewer 2: I think this line needs to be further up, not under the `Close` method details

#### `ConfigValidators` and `ValidateConfig` methods details
There is not much to say here, since this is the same lifecycle that a datasource is having.

#### Testing support

> The testing support will be documented later into a different RFC, or as amendment to this one.

Reviewer 2: Can you give a reason as to why we're not putting it in this RFC?

Author: I didn't dig into this part yet and I barely touched the testing feature of this project.
If you think that this should be added, I can try my best but might take some time until I get a good understanding of the things in there.

Reviewer: Given the limited usage of the testing feature, I agree with @yottta here.  We should focus on the main impact of this feature.

Reviewer 2: I think I'm just looking for some reasoning to be added to the RFC. I'm fine with not doing it now, we should just document why.

Author: Quick fix: dd4ca4729bfb730c406cd06ceaa52e2d6fcfea25

### Support in already ephemeral contexts
There are already OpenTofu contexts that are not saved in state/plan file:
* `provider` configuration
* `provisioner` blocks
* `connection` blocks

In all of these, referencing an ephemeral value should work as normal.

### Utilities
#### `terraform.applying`

> The `terraform.applying` needs to be introduced to allow the user to check if the current command that is running is `apply` or not.

Reviewer 2: why terraform and not tofu?

Author: Because of the feature parity. Maybe having both just to be able to migrate? And in case we encounter `terraform.applying` write a warning that we also support `tofu.applying`? 🤔

Reviewer: Internally, they both map to the same structure for now.  Let's at least mention tofu.applying here.

Author: Done e8c189287f194d9ea8f2319658605ad22d618f45

This is useful when user wants to configure different properties between write operations and read operations.

`terraform.applying` will be set to `true` when `tofu apply` is executed and `false` in any other command.

> [!NOTE]
>

> > This keyword is related to the `apply` command and not to the `apply` phase, meaning that when running `tofu apply`, `terraform.applying` will still be `true` also during the `plan` phase of the `apply` command.

Reviewer 2: what about during destroy?

Author: Fixed in 2c6587907eae9c404931ce789091dfd192f5628b

This is an ephemeral value that should be handled correctly and ensure that its value or any other value generate from it will not end up in a plan/state file.

#### `ephemeralasnull` function

> `ephemeralsnull` function is useful when an object that was built by referencing an ephemeral value wants to be used into a non-ephemeral context.

Reviewer 2: should we recommend that people use this with a nice warning in the UI when they try and create root module outputs with ephemeral values ?

Author: That's a good point 🤔 I like it! Thanks
Added in 2c6587907eae9c404931ce789091dfd192f5628b

  * By using this you don't need to choose what to store and what not, but everything is safely stored.
  * This is also preventing state tempering and [privilege escalation](https://www.plerion.com/blog/hacking-terraform-state-for-privilege-escalation).

## Proposed Solution
Two new concepts will be introduced:
* `ephemeral` resources
* `resource`'s `write-only` attributes

Several existing features will have to be able to be updated with the new functionality:
* variables
* outputs
* locals
* providers
* provisioners
* `connection` block

> [!NOTE]
>
> In order to provide a similar and familiar UX for the users, the proposal in this RFC is heavily inspired by the Terraform public documentation and available blog posts on the matter.

In the attempt to provide to the reader an in-depth understanding of the ephemerality implications in OpenTofu,
this section will try to explain the functional approach of the new concept in each existing feature.

### User Documentation
#### Write-only attributes
This is a new concept that allows any existing `resource` to define attributes in its schema that can be only written without the ability to retrieve the value afterwards.

> By not being readable, this also means that an attribute configured by a provider this way, should not be written to the state or plan file either.

Reviewer 3: Correctness @grammarly
```suggestion
By not being readable, this also means that an attribute configured by a provider this way should not be written to the state or plan file either.
```

Author: Wanted to add the suggestions individually but the signoff of the commit is not working correctly from the GH UI, so I added all your suggestions in one commit: 4ee1ec72ed946b32d50637bb893c4c754d1b6e4e

Therefore, these attributes are suitable for configuring specific resources with sensitive data, like passwords, access keys, etc.

A write-only attribute can accept an ephemeral or a non-ephemeral value, even though it's recommended to use ephemeral values for such attributes.

> Because these attributes are not written to the plan file, the update of a write-only attribute it's getting a little bit trickier.

Reviewer 3: Correctness And Clarity @grammarly
```suggestion
Because these attributes are not written to the plan file, updating a write-only attribute is getting a bit trickier.
```

Author: Wanted to add the suggestions individually but the signoff of the commit is not working correctly from the GH UI, so I added all your suggestions in one commit: 4ee1ec72ed946b32d50637bb893c4c754d1b6e4e

│
│   on main.tf line 21, in module "secret_management":
│   21:   secret_map     = var.secrets
│
│ Variable `secret_map` is not marked as ephemeral. Therefore, it cannot reference an ephemeral value. In case this is actually wanted, you can add the following attribute to its declaration:
│   ephemeral = true
```

OpenTofu should not store ephemeral variable(s) in plan files.
If a plan is generated from a configuration that is having at least one ephemeral variable,
when the planfile will be applied, the value(s) for the ephemeral variable(s) needs to be provided again.

#### Outputs
Most `output` blocks can be configured as ephemeral.

To mark an output as ephemeral, use the following syntax:
```hcl
output "test" {
  // ...
  ephemeral = true
}
```

The ephemeral outputs are available during plan and apply phase and can be accessed only in specific contexts:
* ephemeral variables
* other ephemeral outputs
* write-only attributes
* ephemeral resources
* `provisioner` block
* `connection` block

An `output` block from a root module cannot be marked as ephemeral.
This limitation is natural since ephemeral outputs are meant to be skipped from the state file.

> Therefore, there is no usage of such a defined output block in a root module.

Reviewer 3: Correctness @grammarly
```suggestion
Therefore, there is no use for such a defined output block in a root module.
```

Author: Wanted to add the suggestions individually but the signoff of the commit is not working correctly from the GH UI, so I added all your suggestions in one commit: 4ee1ec72ed946b32d50637bb893c4c754d1b6e4e

> * `connection` block

Author: Added in 5c34b67c8d5e71d22fe4116af6bc9929ea2ee97d

When encountering an ephemeral output in a root module, an error similar to this one should be shown:
```hcl
│ Error: Unallowed ephemeral output
│
│   on main.tf line 36:
│   36: output "write_only_out" {
│
│ Root module is not allowed to have ephemeral outputs
```

Ephemeral outputs are useful when a child module returns sensitive data,
allowing the caller to use the value of that output in other ephemeral contexts.
When using outputs in non-ephemeral contexts, OpenTofu should show an error similar to the following:
```hcl
│ Error: Invalid use of an ephemeral value
│
│   with aws_secretsmanager_secret_version.store_from_ephemeral_output,
│   on main.tf line 31, in resource "aws_secretsmanager_secret_version" "store_from_ephemeral_output":
│   31:   secret_string = module.secret_management.secrets
│

> │ "secret_string" cannot accept an ephemeral value because it is not a write-only attribute which means that will be written to the state.

Reviewer 3: Correctness @grammarly
```suggestion
│ "secret_string" cannot accept an ephemeral value because it is not a write-only attribute, meaning it will be written to the state.
```

* ephemeral variables
* other ephemeral locals
* write-only attributes
* ephemeral resources
* `provider` blocks configuration
* `connection` and `provisioner` blocks

#### Ephemeral resource
In contrast with the write-only arguments where only specifically tagged attributes are not stored in the state/plan file, `ephemeral` resources are not stored entirely.
The ephemeral blocks are behaving similar to `data`, where it reads the indicated resource and once it's done with it, is going to close it.

Ephemeral resources can be referenced only in specific contexts:
* other ephemeral resources
* ephemeral variables
* ephemeral outputs
* locals
* to configure `provider` blocks
* in `provisioner` and `connection` blocks
* in write-only arguments

For example, you can have an ephemeral resource that is retrieving the password from a secret manager, password that can be passed later into a write-only attribute of another normal `resource`.
To do so, the flow of an ephemeral resource should look similar to the following:
* Requests the information from the provider.
* Is passing it into the evaluation context that will be used to evaluate expressions that are referencing it.

>   * Is not storing the value in plan or state file.

Author: Wanted to add the suggestions individually but the signoff of the commit is not working correctly from the GH UI, so I added all your suggestions in one commit: 4ee1ec72ed946b32d50637bb893c4c754d1b6e4e

* When accessing the value, OpenTofu will have to check for the presence of the `RenewAt`:
  * If the timestamp is having a value and the current timestamp is at or over the timestamp indicated by `RenewAt`, call the provider `Renew` method.
* When a reference is found to an ephemeral resource, OpenTofu should double check that the attribute referencing it is allowed to do so.
  * See the contexts above where an ephemeral resource can be referenced.
* At the end, the ephemeral resource needs to be closed. OpenTofu should call `Close` on the provider for all opened ephemeral resources.

> Beside the attributes in a schema of an ephemeral resource, the block should support also the meta-arguments existing in OpenTofu:

Author: Wanted to add the suggestions individually but the signoff of the commit is not working correctly from the GH UI, so I added all your suggestions in one commit: 4ee1ec72ed946b32d50637bb893c4c754d1b6e4e

```bash
ephemeral.playground_random.password: Opening...
ephemeral.playground_random.password: Opening succeeded after 0s
ephemeral.playground_random.password: Closing...
ephemeral.playground_random.password: Closing succeeded after 0s
```

Methods that an ephemeral resource should/could have:
* Required:
  * Open - should be called on the provider to read the information of the indicated resource.
  * Metadata - returns the type name of the ephemeral resource.
  * Schema - should be called on the provider to get the schema defined for the ephemeral resource.
* Optional:
  * Renew - if the response from `Open` contains a valid `RenewAt`, OpenTofu should call this method in order to instruct the provider to renew any possible remote information related to the secret returned from the `Open` call.
  * Close - should be called on the provider to clean any possible remote information related to the secret returned in the response from `Open`.
  * ValidateConfig - validates the configuration provided for the ephemeral resource.

Ephemeral resources lifecycle is similar with the data blocks:

> * Both basic implementations require the same methods (`Metadata` and `Schema`) while datasource is defining `Read` compared with the ephemeral resource that is defining `Open`. When talking about the basic functionality of the ephemeral resources, the `Open` method should behave similarly with the `Read` on a datasource, where it asks the provider for the data associated with that particular ephemeral resource.

Author: Wanted to add the suggestions individually but the signoff of the commit is not working correctly from the GH UI, so I added all your suggestions in one commit: 4ee1ec72ed946b32d50637bb893c4c754d1b6e4e

> the `apply` command. This is `true` also when running a destroy operation.

This is an ephemeral value that should be handled accordingly, meaning that its value or any other value generated 
from it will not end up in a plan/state file.

> [!NOTE]
> 
> For feature parity, the same functionality under `tofu.applying` should be available under `terraform.applying` too.

#### `ephemeralasnull` function
`ephemeralasnull` function is useful when an object built by referencing an ephemeral value wants to be used into a non-ephemeral context.
This is getting a dynamic value and by traversing it, is looking for any ephemeral value and is nullifying it, but it does not nullify any non-ephemeral value within the object.

For example:
```hcl
variable "secret" {
  type = string
  default = "test"
  ephemeral = true
}

locals {
  config = {
    "non-ephemeral": "non-ephemeral-value"
    "ephemeral": var.secret
  }
}

output "test" {
  value = ephemeralasnull(local.config)
}
```

Which after running `tofu apply` should show an output like this:
```hcl
test = {
  "ephemeral" = tostring(null)
  "non-ephemeral" = "non-ephemeral-value"
}
```

> This function should work perfectly fine also with a non-ephemeral value.

Author: Wanted to add the suggestions individually but the signoff of the commit is not working correctly from the GH UI, so I added all your suggestions in one commit: 4ee1ec72ed946b32d50637bb893c4c754d1b6e4e
